src/badguy/icecrusher.cpp

```diff
@@ -220,6 +231,55 @@ IceCrusher::collision_solid(const CollisionHit& hit)
         }
         set_state(RECOVERING);
       }
+      break;
+    case CRUSHING_UP:
+      if (hit.top) {
+        if (ic_size == LARGE) {
+          cooldown_timer = PAUSE_TIME_LARGE;
+          Sector::get().get_camera().shake (0.125f, 0.0f, 16.0f);
+          SoundManager::current()->play("sounds/brick.wav");
+          // throw some particles, bigger and more for large icecrusher
+          for (int j = 0; j < 9; j++)
+          {
+            Sector::get().add<Particles>(
+              Vector(m_col.m_bbox.get_right() - static_cast<float>(j) * 8.0f - 4.0f, m_col.m_bbox.get_top()),
+              0, 90-5*j, 140.0f, 380.0f, Vector(0.0f, 300.0f),
+              1, Color(.6f, .6f, .6f), 5, 1.8f, LAYER_OBJECTS+1);
+            Sector::get().add<Particles>(
+              Vector(m_col.m_bbox.get_left() + static_cast<float>(j) * 8.0f + 4.0f, m_col.m_bbox.get_top()),
+              270+5*j, 360, 140.0f, 380.0f, Vector(0.0f, 300.0f),
+              1, Color(.6f, .6f, .6f), 5, 1.8f, LAYER_OBJECTS+1);
+          }
+        }
+        else {
+          cooldown_timer = PAUSE_TIME_NORMAL;
+          Sector::get().get_camera().shake (0.1f, 0.0, 8.0);
+          if ( m_sprite_name.find("rock_crusher") != std::string::npos ||
+              m_sprite_name.find("moss_crusher") != std::string::npos )
+          {
+            SoundManager::current()->play("sounds/thud.ogg");
+          }
+          else
+          {
+            SoundManager::current()->play("sounds/brick.wav");
+          }
+          // throw some particles
+          for (int j = 0; j < 5; j++)
+          {
+            Sector::get().add<Particles>(
+              Vector(m_col.m_bbox.get_right() - static_cast<float>(j) * 8.0f - 4.0f,
+                     m_col.m_bbox.get_top()),
+              0, 90+10*j, 140, 260, Vector(0, 300),
+              1, Color(.6f, .6f, .6f), 4, 1.6f, LAYER_OBJECTS+1);
+            Sector::get().add<Particles>(
+              Vector(m_col.m_bbox.get_left() + static_cast<float>(j) * 8.0f + 4.0f,
+                     m_col.m_bbox.get_top()),
+              270+10*j, 360, 140, 260, Vector(0, 300),
+              1, Color(.6f, .6f, .6f), 4, 1.6f, LAYER_OBJECTS+1);
+          }
+        }
+        set_state(RECOVERING_UP);
```

Comment:
There is a lot of code duplication here - would it be possible to combine them?

Reply:
I'll try to clean `icecrusher.cpp` a bit (this included), but since Daniel is supposedly working on an icecrusher code revamp, I won't go into any kind of extra ultra mega tidying mode.